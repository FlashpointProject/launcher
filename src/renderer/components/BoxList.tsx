import { useMouse } from '@renderer/hooks/useMouse';
import { findElementAncestor } from '@renderer/Util';
import { OpenIcon } from './OpenIcon';

export type BoxListProps<ItemType> = {
  indexAttr?: string;
  items: ItemType[];
  onRemove: (dataId: number) => void;
  getIndexAttr: (item: ItemType) => number;
  getItemValue: (item: ItemType) => string;
  getColor?: (item: ItemType) => string | undefined;
  renderIcon?: (item: ItemType) => JSX.Element;
}

export function BoxList<ItemType>(props: BoxListProps<ItemType>) {
  const indexAttr = props.indexAttr || 'data-index';
  const { items, getItemValue, getIndexAttr, getColor, onRemove } = props;

  const [onTagMouseDown, onTagMouseUp] = useMouse<number>(() => ({
    chain_delay: 500,
    find_id: (event) => {
      let dataId: number | undefined;
      try { dataId = findAncestorRowDataIndex(event.target as Element, indexAttr); }
      catch (error) { console.error(error); }
      return dataId;
    },
    on_click: (event, dataId, clicks) => {
      if (event.button === 0 && clicks === 1) { // Single left click
        onRemove(dataId);
      }
    },
  }), [props.onRemove]);

  const getIcon = (item: ItemType) => {
    if (props.renderIcon) {
      return props.renderIcon(item);
    } else {
      return (
        <OpenIcon
          className='curate-tag__icon'
          color={getColor ? getColor(item) || '#FFFFFF' : '#FFFFFF'}
          icon='x' />
      );
    }
  };
  return (
    <tr>
      <td/>
      <td
        onMouseDown={onTagMouseDown}
        onMouseUp={onTagMouseUp}>
        { items.length > 0 ? (
          items.map((item, index) => {
            return (
              <div
                className='curate-tag'
                key={index}
                { ...{ [indexAttr]: getIndexAttr(item) } }>
                {getIcon(item)}
                <span className='curate-tag__text'>
                  {getItemValue(item)}
                </span>
              </div>
            );
          })
        ) : undefined }
      </td>
    </tr>
  );
}

function findAncestorRowDataIndex(element: Element, dataIndex: string): number | undefined {
  const ancestor = findElementAncestor(element, target => target.getAttribute(dataIndex) !== null, true);
  if (!ancestor) { return undefined; }

  const index = ancestor.getAttribute(dataIndex);
  if (typeof index !== 'string') { throw new Error('Failed to get attribute from ancestor!'); }

  return (index as any) * 1; // Coerce to number
}
