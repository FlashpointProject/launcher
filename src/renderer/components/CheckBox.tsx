import * as React from 'react';
import { useCallback } from 'react';
import { Omit } from '@shared/interfaces';

/** Props for an input element. */
type InputProps = React.DetailedHTMLProps<React.InputHTMLAttributes<HTMLInputElement>, HTMLInputElement>;

export type CheckBoxProps = Omit<InputProps, 'type'> & {
  /** Called when the checkbox becomes checked or unchecked. This is called right after "onChange". */
  onToggle?: (isChecked: boolean) => void;
};

/** Basic checkbox element. Wrapper around the <input> element. */
export function CheckBox(props: CheckBoxProps) {
  const { onToggle, onChange, ...rest } = props;
  // Hooks
  const onChangeCallback = useCallback((event: React.ChangeEvent<HTMLInputElement>) => {
    if (onChange) { onChange(event); }
    if (onToggle) { onToggle(event.target.checked); }
  }, [onToggle, onChange]);
  // Render
  return (
    <input
      { ...rest }
      type='checkbox'
      onChange={onChangeCallback} />
  );
}
