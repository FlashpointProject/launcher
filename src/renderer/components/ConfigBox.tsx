import * as React from 'react';

export type ConfigBoxProps = {
  title: string;
  description: string;
  contentClassName?: string;
  bottomChildren?: JSX.Element | JSX.Element[];
}

export function ConfigBox(props: React.PropsWithChildren<ConfigBoxProps>) {
  return (
    <div className='setting__row'>
      <div className='setting__row__top'>
        <p className='setting__row__title'>{props.title}</p>
        <div className={`setting__row__content ${props.contentClassName}`}>
          { props.children }
        </div>
      </div>
      <div className='setting__row__bottom'>
        <p>{props.description}</p>
        {props.bottomChildren}
      </div>
    </div>
  );
}
