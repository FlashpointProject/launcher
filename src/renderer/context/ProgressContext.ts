
import { createContextReducer } from '../context-reducer/contextReducer';
import { ReducerAction } from '../context-reducer/interfaces';
import { uuid } from '../util/uuid';

export type ProgressData = {
  /* Key identifier */
  key: string;
  /* Items counted */
  itemCount: number;
  /* Total items to count */
  totalItems: number;
  /* Percent done */
  percentDone: number;
  /* Whether to use percentDone (info for progress components) */
  usePercentDone: boolean;
  /* If bar is finished (now invisible) */
  isDone: boolean;
  /* Text to display (primary) */
  text?: string;
  /* Text to display (secondary) */
  secondaryText?: string;
}

export type ProgressAction = (
  ReducerAction<'set-percent-done', {
    /** Identifier for parent component */
    parentKey: string;
    /** Key progress data to modify. */
    key: string;
    /** Percent done to set */
    percentDone: number;
  }> |
  ReducerAction<'set-use-percent-done', {
    /** Identifier for parent component */
    parentKey: string;
    /** Key progress data to modify. */
    key: string;
    /** Percent done to set */
    usePercentDone: boolean;
  }> |
  ReducerAction<'set-text', {
    /** Identifier for parent component */
    parentKey: string;
    /** Key progress data to modify. */
    key: string;
    /** Percent done to set */
    text: string;
  }> |
  ReducerAction<'set-secondary-text', {
    /** Identifier for parent component */
    parentKey: string;
    /** Key progress data to modify. */
    key: string;
    /** Percent done to set */
    secondaryText: string;
  }> |
  ReducerAction<'set-total-items', {
    /** Identifier for parent component */
    parentKey: string;
    /** Key progress data to modify. */
    key: string;
    /** Percent done to set */
    totalItems: number;
  }> |
  ReducerAction<'count-item', {
    /** Identifier for parent component */
    parentKey: string;
    /** Key progress data to modify. */
    key: string;
  }> |
  ReducerAction<'finished', {
    /** Identifier for parent component */
    parentKey: string;
    /** Key progress data to modify. */
    key: string;
  }>
);

// Empty Record
const defaultProgressState = {};

/** Stores the progress data of various components. */
export const ProgressContext = createContextReducer(
  progressReducer,
  defaultProgressState
);

/** Reducer for the progress state. */
function progressReducer(prevState: Record<string, (ProgressData[]|undefined)>, action: ProgressAction) {
  // Things to keep in mind when writing an action handler:
  // * Don't mutate the previous state or the action object.
  // * Objects/Arrays that have at least one property/element change value should be copied, so that
  //   their reference is different after reducing (this way it is cheap to check which objects/arrays
  //   changed). This should be applied "recursively" so you can trace all the changed values from the
  //   root state object.
  switch (action.type) {
    // Set Percent Done
    case 'set-percent-done': {
      const { parentKey, key } = action.payload;
      const newState = { ...prevState };
      const newParent = ensureParent(newState, parentKey);
      const index = ensureProgressIndex(newParent, key);
      newParent[index].percentDone = action.payload.percentDone;
      newState[parentKey] = newParent;
      return newState;
    }
    // Set suggestion for whether rendering should use percent done
    case 'set-use-percent-done': {
      const { parentKey, key } = action.payload;
      const newState = { ...prevState };
      const newParent = ensureParent(newState, parentKey);
      const index = ensureProgressIndex(newParent, key);
      newParent[index].usePercentDone = action.payload.usePercentDone;
      newState[parentKey] = newParent;
      return newState;
    }
    // Set primary text
    case 'set-text': {
      const { parentKey, key } = action.payload;
      const newState = { ...prevState };
      const newParent = ensureParent(newState, parentKey);
      const index = ensureProgressIndex(newParent, key);
      newParent[index].text = action.payload.text;
      newState[parentKey] = newParent;
      return newState;
    }
    // Set secondary text
    case 'set-secondary-text': {
      const { parentKey, key } = action.payload;
      const newState = { ...prevState };
      const newParent = ensureParent(newState, parentKey);
      const index = ensureProgressIndex(newParent, key);
      newParent[index].secondaryText = action.payload.secondaryText;
      newState[parentKey] = newParent;
      return newState;
    }
    // Set total items
    case 'set-total-items': {
      const { parentKey, key } = action.payload;
      const newState = { ...prevState };
      const newParent = ensureParent(newState, parentKey);
      const index = ensureProgressIndex(newParent, key);
      newParent[index].totalItems = action.payload.totalItems;
      newState[parentKey] = newParent;
      return newState;
    }
    // Count an item and update Percent Done
    case 'count-item': {
      const { parentKey, key } = action.payload;
      const newState = { ...prevState };
      const newParent = ensureParent(newState, parentKey);
      const index = ensureProgressIndex(newParent, key);
      const newItemCount = newParent[index].itemCount += 1;
      const newPercentDone = Math.min((newItemCount / newParent[index].totalItems) * 100, 100);
      newParent[index].itemCount = newItemCount;
      newParent[index].percentDone = newPercentDone;
      newState[parentKey] = newParent;
      return newState;
    }
    // Mark progress as finished by removing it
    case 'finished': {
      const { parentKey, key } = action.payload;
      const newState = { ...prevState };
      const newParent = ensureParent(newState, parentKey);
      const index = newParent.findIndex(data => data ? data.key === key : false);
      if (index >= 0) { newParent.splice(index, 1); }
      newState[parentKey] = newParent;
      return newState;
    }
    default:
      throw Error('ProgressAction sent had an invalid type.');
  }
}

/**
 * Returns the ProgressData of a parent (initializes it if necessary)
 * @param state Mutable ProgressState
 * @param key Parent's Key
 */
function ensureParent(state: Record<string, (ProgressData[]|undefined)>, parentKey: string): ProgressData[] {
  const parent = state[parentKey];
  if (parent) { return [ ...parent ]; }
  else {
    const parent: ProgressData[] = [];
    state[parentKey] = parent;
    return parent;
  }
}

/**
 * Returns the index to the unique ProgressData inside a parent's ProgressData array
 * @param state State of a parent's ProgressData ( state[parentKey] )
 * @param key Unique progress key
 */
function ensureProgressIndex(state: ProgressData[], key: string): number {
  // Return ProgressData for key if exists
  const index = state.findIndex(data => data ? data.key === key : false);
  if (index >= 0) {
    return index;
  } else {
    // Create new data, return index (new length - 1)
    return state.push({
      key: key,
      itemCount: 0,
      totalItems: 0,
      percentDone: 0,
      usePercentDone: true,
      isDone: false
    }) - 1;
  }
}

/**
 * Return a new ProgressHandle to be given out to other functions
 * @param parentKey Identifier of parent (page, component etc.)
 * @param dispatch Dispatcher to ProgressContext
 * @returns Handle containing all data necessary to dispatch ProgressAction's
 */
export function newProgress(parentKey: string, dispatch: React.Dispatch<ProgressAction>): ProgressHandle {
  return {
    parentKey: parentKey,
    key: uuid(),
    dispatch: dispatch
  };
}

/** Passed around to allow recording progress */
export type ProgressHandle = {
  parentKey: string;
  key: string;
  dispatch: React.Dispatch<ProgressAction>;
}


/* Util functions for dispatching Progress Actions given a Progress Handle */
export class ProgressDispatch {
  public static setPercentDone = (handle: ProgressHandle, percentDone: number) => {
    handle.dispatch({
      type: 'set-percent-done',
      payload: {
        parentKey: handle.parentKey,
        key: handle.key,
        percentDone: percentDone
      }
    });
  }

  public static setUsePercentDone = (handle: ProgressHandle, usePercentDone: boolean) => {
    handle.dispatch({
      type: 'set-use-percent-done',
      payload: {
        parentKey: handle.parentKey,
        key: handle.key,
        usePercentDone: usePercentDone
      }
    });
  }

  public static setText = (handle: ProgressHandle, text: string) => {
    handle.dispatch({
      type: 'set-text',
      payload: {
        parentKey: handle.parentKey,
        key: handle.key,
        text: text
      }
    });
  }

  public static setSecondaryText = (handle: ProgressHandle, text: string) => {
    handle.dispatch({
      type: 'set-secondary-text',
      payload: {
        parentKey: handle.parentKey,
        key: handle.key,
        secondaryText: text
      }
    });
  }

  public static setTotalItems = (handle: ProgressHandle, totalItems: number) => {
    handle.dispatch({
      type: 'set-total-items',
      payload: {
        parentKey: handle.parentKey,
        key: handle.key,
        totalItems: totalItems
      }
    });
  }

  public static countItem = (handle: ProgressHandle) => {
    handle.dispatch({
      type: 'count-item',
      payload: {
        parentKey: handle.parentKey,
        key: handle.key
      }
    });
  }

  public static finished = (handle: ProgressHandle) => {
    handle.dispatch({
      type: 'finished',
      payload: {
        parentKey: handle.parentKey,
        key: handle.key
      }
    });
  }
}
