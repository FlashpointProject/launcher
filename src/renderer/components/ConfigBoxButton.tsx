import * as React from 'react';
import { ConfigBox, ConfigBoxProps } from './ConfigBox';
import { SimpleButton, SimpleButtonProps } from './SimpleButton';

export type ConfigBoxButtonProps = ConfigBoxProps & SimpleButtonProps;

export function ConfigBoxButton(props: ConfigBoxButtonProps) {
  return (
    <ConfigBox
      {...props}
      contentClassName={`${props.contentClassName || ''} setting__row__content--button`}>
      <div>
        <SimpleButton {...props} />
      </div>
    </ConfigBox>
  );
}
