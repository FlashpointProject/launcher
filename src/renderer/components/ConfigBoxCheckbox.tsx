import * as React from 'react';
import { CheckBox, CheckBoxProps } from './CheckBox';
import { ConfigBox, ConfigBoxProps } from './ConfigBox';

export type ConfigBoxCheckboxProps = ConfigBoxProps & CheckBoxProps;

export function ConfigBoxCheckbox(props: ConfigBoxCheckboxProps) {
  return (
    <ConfigBox
      {...props}
      contentClassName={`${props.contentClassName || ''} setting__row__content--toggle`}>
      <div>
        <CheckBox {...props}/>
      </div>
    </ConfigBox>
  );
}
