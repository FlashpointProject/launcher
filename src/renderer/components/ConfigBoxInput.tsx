import * as React from 'react';
import { ConfigBox, ConfigBoxProps } from './ConfigBox';
import { InputField, InputFieldProps } from './InputField';

export type ConfigBoxInputProps = ConfigBoxProps & InputFieldProps;

export function ConfigBoxInput(props: ConfigBoxInputProps) {
  return (
    <ConfigBox {...props}>
      <InputField {...props}/>
    </ConfigBox>
  );
}
