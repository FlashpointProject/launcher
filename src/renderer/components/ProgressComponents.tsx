import * as React from 'react';
import { ProgressData } from '../context/ProgressContext';
import { LangContext } from '../util/lang';

export type ProgressComponentProps = {
  /** Data to read from. */
  progressData: ProgressData;
  /** Class name to append to the wrapper. */
  wrapperClass?: string;
}

/** Choose a StatusBar or ProgressBar, letting the Progress handler decide. */
export function AutoProgressComponent(props: ProgressComponentProps) {
  if (!props.progressData.isDone) {
    if (props.progressData.usePercentDone) {
      return (
        <ProgressBar
          progressData={props.progressData}
          wrapperClass={props.wrapperClass} />
      );
    } else {
      return (
        <StatusBar
          progressData={props.progressData}
          wrapperClass={props.wrapperClass} />
      );
    }
  } else {
    return (<></>);
  }
}

/** Large centered primary text, smaller underneath secondary text. */
export function StatusBar(props: ProgressComponentProps) {
  return (
    <div className={(props.wrapperClass ? ' ' + props.wrapperClass : '') + 'progress-component__wrapper'}>
      <div className='status-bar__top-text'>{props.progressData.text}</div>
      { props.progressData.secondaryText ? (
        <div className='status-bar__bottom-text'>
          {props.progressData.secondaryText}
        </div>
      ) : undefined }
    </div>
  );
}

/** Large top text `Percent% Complete`, medium progress bar, small underneath primary text. */
export function ProgressBar(props: ProgressComponentProps) {
  const strings = React.useContext(LangContext);
  const barCssProps: React.CSSProperties = React.useMemo(() => ({
    width: `${props.progressData.percentDone}%`
  }), [props.progressData.percentDone]);

  return (
    <div className={'progress-component__wrapper' + (props.wrapperClass ? ' ' + props.wrapperClass : '')}>
      <div className='progress-bar__top-text'>{`${Math.round(props.progressData.percentDone)}% ${strings.misc.complete}`}</div>
      <div className='progress-bar__bar'>
        <div className='progress-bar__bar__inner' style={barCssProps} />
      </div>
      { props.progressData.text ? (
        <div className='progress-bar__bottom-text'>
          {props.progressData.text}
        </div>
      ) : undefined }
    </div>
  );
}
