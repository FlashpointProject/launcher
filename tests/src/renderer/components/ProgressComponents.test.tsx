import { configure, shallow, mount } from 'enzyme';
import * as Adapter from 'enzyme-adapter-react-16';
import * as React from 'react';
import { AutoProgressComponent, StatusBar, ProgressBar } from '@renderer/components/ProgressComponents';
import { ProgressData } from '@renderer/context/ProgressContext';
import toJson from 'enzyme-to-json';

configure({ adapter: new Adapter() });

describe('Progress Components', () => {
  test('Auto Select Component', () => {
    const mockDataStatus = createMockData();
    const componentStatus = shallow(
      <AutoProgressComponent progressData={mockDataStatus} wrapperClass={'TestSuccess'} />
    );
    expect(toJson(componentStatus)).toMatchSnapshot();

    const mockDataProgress = {...createMockData(), usePercentDone: true};
    const componentProgress = shallow(
      <AutoProgressComponent progressData={mockDataProgress} wrapperClass={'TestSuccess'} />
    );
    expect(toJson(componentProgress)).toMatchSnapshot();
  });

  test('Stop Render When Done', () => {
    const mockDataDone = {...createMockData(), isDone: true};
    const component = shallow(
      <AutoProgressComponent progressData={mockDataDone} wrapperClass={'TestSuccess'} />
    );
    expect(toJson(component)).toMatchSnapshot();
  });

  test('Status Bar Render', () => {
    const mockData = createMockData();
    const component = mount(
      <StatusBar progressData={mockData} wrapperClass={'TestSuccess'} />
    );
    expect(toJson(component)).toMatchSnapshot();
  });

  test('Progress Bar Render', () => {
    const mockData = createMockData();
    const component = mount(
      <ProgressBar progressData={mockData} wrapperClass={'TestSuccess'} />
    );
    expect(toJson(component)).toMatchSnapshot();
  });
});

function createMockData(): ProgressData {
  return {
    key: 'test',
    percentDone: 0,
    usePercentDone: false,
    isDone: false,
    text: 'TestSuccess',
    secondaryText: 'TestSuccess',
    itemCount: 0,
    totalItems: 0
  };
}