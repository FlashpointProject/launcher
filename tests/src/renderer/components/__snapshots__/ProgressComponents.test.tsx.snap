// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Progress Components Auto Select Component 1`] = `
<StatusBar
  progressData={
    Object {
      "isDone": false,
      "itemCount": 0,
      "key": "test",
      "percentDone": 0,
      "secondaryText": "TestSuccess",
      "text": "TestSuccess",
      "totalItems": 0,
      "usePercentDone": false,
    }
  }
  wrapperClass="TestSuccess"
/>
`;

exports[`Progress Components Auto Select Component 2`] = `
<ProgressBar
  progressData={
    Object {
      "isDone": false,
      "itemCount": 0,
      "key": "test",
      "percentDone": 0,
      "secondaryText": "TestSuccess",
      "text": "TestSuccess",
      "totalItems": 0,
      "usePercentDone": true,
    }
  }
  wrapperClass="TestSuccess"
/>
`;

exports[`Progress Components Progress Bar Render 1`] = `
<ProgressBar
  progressData={
    Object {
      "isDone": false,
      "itemCount": 0,
      "key": "test",
      "percentDone": 0,
      "secondaryText": "TestSuccess",
      "text": "TestSuccess",
      "totalItems": 0,
      "usePercentDone": false,
    }
  }
  wrapperClass="TestSuccess"
>
  <div
    className="progress-component__wrapper TestSuccess"
  >
    <div
      className="progress-bar__top-text"
    >
      0% complete
    </div>
    <div
      className="progress-bar__bar"
    >
      <div
        className="progress-bar__bar__inner"
        style={
          Object {
            "width": "0%",
          }
        }
      />
    </div>
    <div
      className="progress-bar__bottom-text"
    >
      TestSuccess
    </div>
  </div>
</ProgressBar>
`;

exports[`Progress Components Status Bar Render 1`] = `
<StatusBar
  progressData={
    Object {
      "isDone": false,
      "itemCount": 0,
      "key": "test",
      "percentDone": 0,
      "secondaryText": "TestSuccess",
      "text": "TestSuccess",
      "totalItems": 0,
      "usePercentDone": false,
    }
  }
  wrapperClass="TestSuccess"
>
  <div
    className=" TestSuccessprogress-component__wrapper"
  >
    <div
      className="status-bar__top-text"
    >
      TestSuccess
    </div>
    <div
      className="status-bar__bottom-text"
    >
      TestSuccess
    </div>
  </div>
</StatusBar>
`;

exports[`Progress Components Stop Render When Done 1`] = `<Fragment />`;
